### ISSUE TITLE
[BUG]: useStorageState hook type error

### ISSUE THREAD

**Issue by ojj1123** [toss/slash#220]
### **Package Scope**

Package name: @toss/react

#### tl;dr
- In the example of `useStorageState`, a type error occurs. This is because `defaultValue` is inferred as a `literal type` rather than a `primitive type`.
- I tried to infer the type of `defaultValue` without writing a `generic` in `useStorageState`, but I couldn't find a way. Please advise if there is any other way
- If there is no other way, I will add a phrase to the document that a `generic` must be created if the `defaultValue` is a `primitive type` (string, boolean, number).

---

This is an issue about a type error occurring in the `useStorageState` hook.

If you try to use the example in [the document of `useStorageState` hook](https://slash.page/libraries/react/react/src/hooks/usestoragestate/) as it is, the following type error occurs.
```ts
function MyComponent() {
  const [state, setState, refresh] = useStorageState('@service/some-resource', {
    defaultValue: 0,
  });

  useEffect(() => {
    setState(x => x + 1);
    // ^ type error
  }, []);

  ...
}
```
<img width="513" alt="type-error" src="https://user-images.githubusercontent.com/33178048/222148878-8f7e8a6a-18c5-407e-9224-cc22025a3987.png">

However, if you write a `generic`, type errors do not occur:
```ts
const [state, setState, refresh] = useStorageState<number>('@service/some-resource', {
  defaultValue: 0,
});

useEffect(() => {
  // No error
  setState((x) => x + 1);
}, [setState]);
```

I confirmed that the `state` among the return values of `useStorageState` in the example above is not inferred as `number`, but as literal type `0`. In addition, even when the `defaultValue` is a primitive type such as `boolean` or `string`, it is inferred only as a `literal type`:
```ts
  // const state1: true
  const [state1, setState1, refresh1] = useStorageState('boolean', {
    defaultValue: true,
  });

  // const state2: "str"
  const [state2, setState2, refresh2] = useStorageState('string', {
    defaultValue: 'str',
  });
```

**Here, I want to make type inference without writing generic like `React.useState`.** However, because of the type constraint that prevents the `symbol` value from entering the `devalutValue` of the `useStorageState` argument, type inference is not possible as a primitive type like `React.useState`.
```ts
// const count: number
const [count, setCount] = useState(1);

// const state: number  ✅ What I want
// const state: 0  😢 Actually
const [state, setState, refresh] = useStorageState('number', {
  defaultValue: 0
})
```

In https://github.com/Microsoft/TypeScript/pull/10676, when inferring function arguments, type parameters are inferred as widened literal types only in the following situations:

> During type argument inference for a call expression the type inferred for a type parameter T is widened to its widened literal type if:
>
> - all inferences for T were made to top-level occurrences of T within the particular parameter type, and
> - **T has no constraint or its constraint does not include primitive or literal types, and**
> - T was fixed during inference or T does not occur at top-level in the return type.

In other words, since the primitive type is included in the type constraint of `useStorageState`, the type inferred from the type parameter is a literal type:(Please check my [playground](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcDKwoEsCGAbLALxwCM9gAeAFQD44BeOKuUGYAOwBMBnObmbOwDmcAD5x2AVwC2JTGLgkIEcjnYLJ7ANbsIAd3YBtALoKASsADG0ThX6ChAGjiad+9nQD8TOAC4JwABumADcAFBhAGaaljBYEOqRytQ0ABRgOFA40v4Y2PhEpOQpAJT+gRBYnHAA3mFwcNbs3CrAAHR4EELpmdkl4QC+EUkQqQBEABbAeJ1j-WEA9AsAep5RyqkALABM80ur66MCksB7K55AA))

https://github.com/toss/slash/blob/main/packages/react/react/src/hooks/useStorageState.ts#L5

```ts
// useStorageState.ts
// A conditional type constraint contains a primitive type
export type Serializable<T> = T extends string | number | boolean | unknown[] | Record<string, unknown> ? T : never;
...
```

**Therefore, I concluded that the type parameter `T` cannot be inferred as a primitive type (e.g. `string`, `number` or `boolean`) because of the type constraint for `defaultValue` in `useStorageState`.** Is there any other way?

If there is no other way, it seems that the phrase that a `generic` must be created when a primitive type enters the `defaultValue` should be added to the official document.

Please give me some advice on the issue and I'll fix the documentation.

**Comment by Jinho1011** [comment]
This issue is resolved by #223

**Comment by Jinho1011** [comment]
I left a comment on #223 , but I think you didn't see it, so I'm leaving it here.

I also confirmed that this PR did not solve the issue. I apologize for my mistake. But I found that the method below widens the type, but what is your opinion on its appearance, even if it is not pretty?

````typescript
type ToPrimitive<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends unknown[]
  ? unknown[]
  : T extends Record<string, unknown>
  ? Record<string, unknown>
  : T;
````

**Comment by ojj1123** [comment]
I'm sorry I'm late checking comments

I tried the solution you gave, but the `primitive type` is inferred well, but **it does not prevent `symbol` from being passed to `defaultValue` as before with a type error.**

If you solved it, could you reproduce it?

> what is your opinion on its appearance, even if it is not pretty?

I agree, but I think this part should be reviewed by the maintainers of this library!

**Comment by Jinho1011** [comment]
<img width="952" alt="image" src="https://user-images.githubusercontent.com/26860466/230609422-2463be03-2b80-4c86-922c-10d9384604fa.png">

It looks like checking for the symbol type too. While limiting the type this way might not be the best approach, it does solve the problem. So, I'll submit a new pull request and wait for the maintainer's feedback, as you suggested.

**Comment by ojj1123** [comment]
This issue is resoloved in this PR #292
